### ISSUE TITLE
[BUG] SwitchMLP is not compatible with distributed_optimizer, but there is no assertion.

### ISSUE THREAD

**Issue by huangyf530** [NVIDIA/Megatron-LM#565]
**Describe the bug**
SwitchMLP used with distributed_optimizer will cause parameters in SwitchMLP not updated by optimizer. The **main_grad of parameters in SwitchMLP is stored in discreate buffers which are not in model.grad_buffers**, and therefore they will not be used in distributed_optimizer and will never update during training process.

**To Reproduce**
Add following arguments in training script:
```shell
--bf16 \
--num-experts 4 \
--num-expert-layers 2 \
--expert-model-parallel-size 2 \
--use-distributed-optimizer \
```
Then you can compare parameters of SwitchNLP among checkpoints for different iterations.

**Expected behavior**
The paramsters should be different among different iterations.

**Stack trace/logs**
N/A

**Environment (please complete the following information):**
 - Megatron-LM commit ID
 - PyTorch version
 - CUDA version
 - NCCL version

**Proposed fix**
1. Maybe we can add parameters of SwichMLP into model's grad_buffers but in different buckets with other parameters. And when all_reduce/reduce_scatter model's gradients, the used parallel_group should be different for different buckets (data_parallel_group for dense parameters and data_modulo_expert_parallel_group for parameters in expert). I can imagine this will still nedd other modifications on distributed_optimizer to make it work. And I am very happy to make my contirbutions to solve this problem.
2. Another simple method may disable distributed_optimizer when SwitchMLP is used.

**Additional context**
N/A

**Comment by yanring** [comment]
@huangyf530 Thank for reporting the issue. Could you please provide the parallel setting in your case?

**Comment by huangyf530** [comment]
@yanring I run it on a toy model with llama's tokenizer on 8 A800 GPU.  Some related arguments are in the following:
```shell
DISTRIBUTED_ARGS=" \
       --tensor-model-parallel-size 2 \
       --pipeline-model-parallel-size 1 \
       --distributed-backend nccl \
       --use-distributed-optimizer \
       "

NETWORK_SIZE_ARGS=" \
       --num-layers 4 \
       --hidden-size 128 \
       --num-attention-heads 4 \
       --group-query-attention \
       --num-query-groups 4 \
       --ffn-hidden-size 512 \
       --position-embedding-type rope \
       --max-position-embeddings 4096 \
       --make-vocab-size-divisible-by 1 \
       --norm-epsilon 1e-5 \
       --normalization RMSNorm \
       --swiglu \
       --untie-embeddings-and-output-weights \
       --use-flash-attn \
       --sequence-parallel \
       "

REGULATIZATION_ARGS=" \
       --attention-dropout 0.1 \
       --hidden-dropout 0.1 \
       --weight-decay 1e-1 \
       --clip-grad 1.0 \
       --adam-beta1 0.9 \
       --adam-beta2 0.95 \
       --adam-eps 1e-8 \
       "

TRAINING_ARGS=" \
       --micro-batch-size 1 \
       --global-batch-size 32 \
       --train-iters 10000 \
       --log-interval 1 \
       --disable-bias-linear \
       --no-bias-gelu-fusion \
       --optimizer adam \
       --exit-interval 10 \
       "

MIXED_PRECISION_ARGS=" \
       --bf16 \
       --no-query-key-layer-scaling \
       "

MOE_ARGS="
       --num-experts 4 \
       --expert-model-parallel-size 2 \
       "
```

**Comment by yanring** [comment]
Thanks for the response! 
I believe the issue lies in Megatron does not support DP for experts at present (DP size should <= EP size). In your situation, you can utilize the command `--expert-model-parallel-size 4` to run the model.

**Comment by huangyf530** [comment]
@yanring Thank you for your response!
I have tried to utilize `--expert-model-parallel-size 4`.  But the parameters of SwithMLP still don't update when I use `--distributed-optimizer`. And when I remove `--distributed-optimizer`, the parameters can be updated. I use the following code to compare experts in different checkpoints.
```python
import torch

filepath1 = "checkpoint/debug_moe/iter_0000005/mp_rank_00_000/model_optim_rng.pt"
filepath2 = "checkpoint/debug_moe/iter_0000010/mp_rank_00_000/model_optim_rng.pt"

model1 = torch.load(filepath1, map_location='cpu')['model']['language_model']['encoder']
model2 = torch.load(filepath2, map_location='cpu')['model']['language_model']['encoder']
print(model1.keys())

# compare router
router1 = model1['layers.0.mlp.router.weight']
router2 = model2['layers.0.mlp.router.weight']
print("layer 0 router")
print(torch.max(torch.abs(router1 - router2)))

# compare expert
expert1 = model1['layers.0.mlp.local_experts.0.dense_h_to_4h.weight']
expert2 = model2['layers.0.mlp.local_experts.0.dense_h_to_4h.weight']
print("layer 0 expert")
print(torch.max(torch.abs(expert1 - expert2)))

# attention dense
mlp1 = model1['layers.0.self_attention.dense.weight']
mlp2 = model2['layers.0.self_attention.dense.weight']
print("layer 0 attention dense")
print(torch.max(torch.abs(mlp1 - mlp2)))

# attention qkv
mlp1 = model1['layers.0.self_attention.query_key_value.weight']
mlp2 = model2['layers.0.self_attention.query_key_value.weight']
print("layer 0 attention qkv")
print(torch.max(torch.abs(mlp1 - mlp2)))
```

With `--distributed-optimizer`, I will get the following output:
```
layer 0 router
tensor(1.9073e-06, dtype=torch.bfloat16)
layer 0 expert
tensor(0., dtype=torch.bfloat16)
layer 0 attention dense
tensor(1.9073e-06, dtype=torch.bfloat16)
layer 0 attention qkv
tensor(1.9073e-06, dtype=torch.bfloat16)
```
Without `--distributed-optimizer` will get the following output:
```
layer 0 router
tensor(1.9073e-06, dtype=torch.bfloat16)
layer 0 expert
tensor(1.9073e-06, dtype=torch.bfloat16)
layer 0 attention dense
tensor(1.9073e-06, dtype=torch.bfloat16)
layer 0 attention qkv
tensor(1.9073e-06, dtype=torch.bfloat16)
```

**Comment by ShinoharaHare** [comment]
Is this still a problem?
In the newly released documentation, I found that the example for training Mixtral uses the following parameters:
```
--expert-model-parallel-size
--use-distributed-optimizer
```
If it remains an issue, it could be quite misleading.

**Comment by ShinoharaHare** [comment]
Is this still a problem?
In the newly released documentation, I found that the [example](https://docs.nvidia.com/megatron-core/developer-guide/latest/api-guide/moe.html#a-detailed-moe-script) for training Mixtral uses the following parameters:
```
--expert-model-parallel-size
--use-distributed-optimizer
```
If it remains an issue, it could be quite misleading.

**Comment by yanring** [comment]
> Is this still a problem? In the newly released documentation, I found that the [example](https://docs.nvidia.com/megatron-core/developer-guide/latest/api-guide/moe.html#a-detailed-moe-script) for training Mixtral uses the following parameters:
> 
> ```
> --expert-model-parallel-size
> --use-distributed-optimizer
> ```
> 
> If it remains an issue, it could be quite misleading.

Thank you for the reminder, DistOpt should now works well with MoE, including Expert Parallelism.

**Comment by ShinoharaHare** [comment]
> > Is this still a problem? In the newly released documentation, I found that the [example](https://docs.nvidia.com/megatron-core/developer-guide/latest/api-guide/moe.html#a-detailed-moe-script) for training Mixtral uses the following parameters:
> > ```
> > --expert-model-parallel-size
> > --use-distributed-optimizer
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If it remains an issue, it could be quite misleading.
> 
> Thank you for the reminder, DistOpt should now works well with MoE, including Expert Parallelism.

Oh, I see. Thank you for clarifying.
Could I also ask why this issue and the related issues haven't been closed yet?
